### PR TITLE
Add matrix logm

### PR DESCRIPTION
**Pull request by MarekPokropinski**
This PR is implementation of matrix logarithm (jax version of `scipy.linalg.logm`). Makes progress on issue #2478. Uses algorithm based on Schur decomposition and inverse squaring and scaling algorithm.

Method is fully described in paper: Awad H. Al-Mohy and Nicholas J. Higham (2012) “Improved Inverse Scaling and Squaring Algorithms for the Matrix Logarithm.” SIAM Journal on Scientific Computing, 34 (4). C152-C169. ISSN 1095-7197

Unlike in scipy in jax implemented function always uses complex types because it is not possible to statically know if result is real or complex, as there is no real solution for matrices with negative eigenvalues.

Used algorithms use a lot of loops and goto's in papers' pseudocodes so code uses a lot of lax control flow operations.

As for performance compared to SciPy:
Figure_2
y axis is time in seconds, and x is matrix size n (of square matrix n x n). Performance excludes first function pass in jax.

### CODE REVIEW

**Comment by dfm** [comment]
@MarekPokropinski — Thanks for this contribution! I'd like to call in @jakevdp's feedback first as to how this fits within [JAX's scope](https://jax.readthedocs.io/en/latest/jep/18137-numpy-scipy-scope.html). While `scipy.linalg` is considered within JAX's scope, this implementation is complicated and (as you allude to in your comment) it isn't particularly well-aligned with JAX or XLA's computation model, so I think there might be some push back to including it in JAX directly. All that being said, this is a really great start, and it would be awesome to make it available it somewhere (either here or ... TBD?)!

Depending on Jake's feedback, here are some high level comments from my initial look:

1. Since this is a big change, it might be better to split this into 2 PRs: one for the Legendre roots, and one for `expm`.
2. Do you have a sense for why the JAX and scipy implementations seem to have different scalings? What is the expected scaling for this algorithm?
3. How does performance compare on GPU and/or TPU? Methods that require significant control flow like this tend to behave poorly on accelerators.
4. Numerics: what data type are you using in your benchmark? How sensitive are the results to 32 vs. 64 bit precision? Also: since this is a randomized algorithm, how sensitive are the results to changes in seed?

**Comment by MarekPokropinski** [comment]
as for 2, I'm not sure, the only thing that comes to my mind is that scipy uses dynamic typing: it uses complex matrices only if it is needed. I used complex numbers always to make static typing easier (now that I think about it in jax it should be possible to do this check too). 
3. not working on gpu nor tpu. I implemented schur based algorithm (schur decomposition is only on cpu in jax for now), the same that is in scipy. I guess for gpu/tpu different implementation could be used, but it would be slower on CPU (more multiplications).
4. I tested only for float32. As for randomization - `_onenormest` function, which is 1-norm estimation, the only function that is random, is sensitive to seed, as is the same function in scipy, but it does not influence overall inverse squaring and scaling algorithm. It is used there to dynamically select s and m parameters for pade approximation, and little mistakes don't affect results that much.

**Comment by MarekPokropinski** [comment]
I did some testing and the lack of performance scaling relative to scipy seems to come from difference in sqrtm function. Scipy uses sylvester based implementation for sqrtm which is faster, but for some reason even if i force scipy to use blocksize=1 implementation in jax is slower.

In pr #9544 author showed that with blocksize set to 1 implementation is faster in jax but it is not on my computer.

**Comment by evan-palmer** [comment]
I tested this out on a GPU with CUDA 12 using a simple implementation of the QR algorithm for Schur decomposition. I limited the test matrices to real matrices to simplify the Schur decomposition implementation. I haven't plotted the results, but the implementation does seem to be slower on GPU than on CPU. For example, a random 5x5 matrix takes ~0.02s to run vs ~0.0015s on CPU (per the graph above). I will note that the compile time is significant on my GPU. During my tests, the initial compilation takes ~30s. I also tried vectorizing the implementation and ultimately killed the process after 10 minutes of compilation.

Below is my QR algorithm implementation for folks that want to test on their own hardware:

```python
@partial(jit, static_argnames=("itmax",))
def schur(T, itmax = 100):
    n = T.shape[0]
    I = jnp.eye(n)

    def make_step(i, args):
        A, Q_t = args

        shift = A[-1, -1] * I
        Q, R = jnp.linalg.qr(A - shift)
        A = R @ Q + shift
        Q_t = Q_t @ Q

        return A, Q_t

    return jax.lax.fori_loop(0, itmax, make_step, (T, I))
```

**Comment by MarekPokropinski** [comment]
@evan-palmer Hey, I did some tests, and this behavior is also in function `sqrtm` that `logm` heavily relies upon. So the dropoff in speed could just come from dropoff in speed of `sqrtm`

**Comment by abhijeetgangan** [comment]
Is there any update on this? I would like to use the logM for the geometry / structural optimization in [`jax-md`](https://github.com/jax-md/jax-md.git). The [algorithm requires](https://gitlab.com/ase/ase/-/blob/master/ase/filters.py#L624) logM implementation at one of the steps. Previously, for torch version I had used a special version for [3D](https://github.com/abhijeetgangan/torch_matfunc/blob/cf5582e4d6b80f8011a537352c884cf2a980d2b3/torch_matfunc/matrix/_logm_33.py#L388) but it would be nice to have it in main given the progress made on this PR and making it work for any matrix size.

**Comment by UgaTheDev** [comment]
Thanks for sticking with this — `logm` is a genuinely wanted gap in `jax.scipy.linalg`, and getting inverse scaling-and-squaring right in a traceable, `jit`-able form is hard work. The overall structure follows Al-Mohy & Higham (2012) Algorithm 4.1 closely and mirrors the existing `_sqrtm` Schur/transform-back pattern, so the shape of this is right. A few things below are load-bearing before it can land.

**1. The diagonal replacement is silently discarded.**
```python
U2 = jnp.fill_diagonal(U, jnp.log(diag), inplace=False)
...
U2 = lax.fori_loop(0, n - 1, replace_superdiag_fn, U)   # starts from U, not U2
```
The second assignment rebinds `U2` from `U`, so `log(diag)` never reaches the result. That's the accuracy fix the principal-branch path exists for, and it's dropped with no error — the function still returns a plausible answer, just a less accurate one. Threading `U2` into the `fori_loop` is the fix, and a test asserting `diag(logm(T)) == log(diag(T))` for a triangular `T` with well-separated eigenvalues would catch it.

**2. The scaling loop can't terminate on a singular input.**
```python
cond = lambda x: jnp.max(jnp.abs(x[0] - 1)) > theta[7]
body = lambda x: (jnp.sqrt(x[0]), x[1] + 1)
```
For an eigenvalue of exactly `0`, `sqrt(0) == 0` and `|0 - 1| == 1 > theta[7]` forever: the `while_loop` hangs on device with no diagnostic. `scipy.linalg.logm` raises for a singular matrix. Suggest an explicit iteration cap plus returning `nan` when the cap is hit, so the failure is visible. A NaN eigenvalue takes the opposite path — the comparison is `False`, the loop exits immediately, and the garbage propagates unflagged.

**3. Two complex-input issues in `_onenormest`.**
`S = (Y + (Y == 0)) / jnp.abs(Y)` intends `sign(Y)`, but where `Y == 0` it evaluates `1 / 0 -> inf` rather than `1`; `jnp.where(Y == 0, 1, Y / jnp.abs(Y))` gives the intended value. Separately, `Z = A.T @ S` should be the conjugate transpose (`A.conj().T`) for the complex case — scipy's estimator uses `A.H`. Both are quiet: they produce a wrong norm estimate, which perturbs the Padé degree `m` rather than raising.

**4. `logm`'s signature diverges from scipy — probably the main thing to settle before more numerics work.**
`logm(A, key)` makes a required PRNG key part of a public API whose scipy counterpart is `logm(A)`, and it makes every call site randomized. The only consumer of `key` is the starting matrix in `_onenormest`. Options: default `key=None` and derive a fixed internal seed (scipy is randomized internally too, it just doesn't surface it), or use the exact 1-norm for small `n` and reserve the estimator for large ones. Also, `sqrtm` on current main routes through `ensure_arraylike(...)` and documents the CPU-only `schur` dependency — `logm` should match both, and the docstring should document `key` under `Args`.

**5. The doctest as written will fail CI.** The example binds `A = jnp.array(...)` then calls `logm(a, ...)` with a lowercase `a` that is never defined. The prose "matrix multiplication is inverse of exponentiation" also looks like it wants to say the matrix logarithm inverts the matrix exponential.

**6. Scope: `roots_legendre` is a second public API addition riding along.** It's genuinely needed for the Gauss-Legendre evaluation of `r_m`, so I'm not arguing to drop it — but its `max_n` parameter deviates from `scipy.special.roots_legendre` in a way callers have to understand, and its docstring says "JAX implementation of `scipy.linalg.logm`" (copy-paste). Neither `logm` nor `roots_legendre` is listed in `docs/jax.scipy.rst`, and there's no CHANGELOG entry — both are needed for a public API addition. If reviewers would rather see this land in two pieces, splitting `roots_legendre` into its own PR would make this one easier to approve.

**7. Tests.** Dtype coverage and the CPU guard look right. Three suggestions: the test matrix construction `mat = arg @ arg.T` isn't Hermitian PSD for complex dtypes — `arg @ arg.conj().T` is what makes the "principal log is unique" premise hold there. `testOneNormEstimator` and `testInverseSquaring` import scipy private internals (`scipy.sparse.linalg._onenormest`, `scipy.linalg._matfuncs_inv_ssq`) at module level, which will break the whole test module on any scipy release that reorganizes those paths — testing public behavior instead would be more durable. And the edge cases most likely to expose items 1–3 aren't covered: identity, a singular matrix, a defective matrix (e.g. a Jordan block), and a triangular matrix with repeated eigenvalues exercising the `l1 == l2` branch.

**Overall:** the algorithm structure and the Schur/transform-back path look sound and consistent with `_sqrtm`. Items 1–3 are correctness bugs that fail quietly rather than loudly, and item 4 is an API question worth settling with a maintainer before more numerics work goes in. The branch also needs a rebase — `sqrtm`'s surrounding context has changed on main. Happy to look again once those are addressed.
